nucleus/payara-modules/payara-executor-service/src/main/java/fish/payara/nucleus/executorservice/PayaraFileWatcher.java

```diff
@@ -0,0 +1,225 @@
+        try {
+            // Block, waiting for an event on a watched file
+            WatchKey key = watcher.take();
+
+            // Loop through the events
+            for (WatchEvent<?> event : key.pollEvents()) {
+                // Find the absolute path of the modified file
+                final Path modifiedPath = ((Path) key.watchable()).resolve((Path) event.context());
+
+                // Loop through the watched paths to find the action associated with it
+                Iterator<Entry<Path, Runnable>> watchIterator = LISTENER_MAP.entrySet().iterator();
+                while (watchIterator.hasNext()) {
+                    Entry<Path, Runnable> watchEntry = watchIterator.next();
+
+                    // If this entry corresponds to the modified file
+                    if (modifiedPath.endsWith(watchEntry.getKey())) {
+                        // Ignore overflow events
+                        if (event.kind() == StandardWatchEventKinds.OVERFLOW) continue;
+
+                        File modifiedFile = modifiedPath.toFile();
+
+                        // If it's been deleted, remove the file watcher
+                        if (event.kind() == StandardWatchEventKinds.ENTRY_DELETE && !modifiedFile.exists()) {
+                            LOGGER.info(format("Watched file %s was deleted; removing the file watcher", modifiedPath));
+                            watchIterator.remove();
+                        } else if (event.kind() == StandardWatchEventKinds.ENTRY_MODIFY && modifiedFile.length() > 0) {
+                            // Run the associated action
+                            LOGGER.fine(format("Watched file %s modified, running the listener", modifiedPath));
+                            watchEntry.getValue().run();
+                        }
+                    }
+                }
+            }
+            key.reset();
+        } catch (InterruptedException ex) {
+            LOGGER.log(WARNING, "The file watcher thread was interrupted", ex);
+        }
+    }
+
+    private synchronized void initialise() {
+        if (!running && !PATHS_TO_WATCH.isEmpty()) {
+            try {
+                watcher = FileSystems.getDefault().newWatchService();
+                executor.scheduleWithFixedDelay(this::run, 0, 1, SECONDS);
+                running = true;
+                LOGGER.info("Initialised the file watcher service");
+            } catch (IOException ex) {
+                LOGGER.log(WARNING, "Failed to initialise the watch service", ex);
+            }
+        }
+    }
+
+
+    private synchronized void terminate() {
+        if (running) {
+            try {
+                watcher.close();
+                running = false;
+                LOGGER.info("Terminated the file watcher service");
+            } catch (IOException ex) {
+                LOGGER.log(WARNING, "Failed to terminate the watch service", ex);
+            }
+        }
+    }
+
+    // Lifecycle methods
```

Comment:
Could these not be at the top?
This is edging dangerously close to the mindset of class variables being declared at the bottom, and that's an abyss that's swallowed many a good developer.

Seriously though they're at the top in most other classes.

Reply:
I haven't seen it explicitly declared in any of our code styling comments so I assumed it was open to interpretation! To me, these methods belong at the back, but I'm happy to move them to the top

Reply:
You're right they don't seem to be mentioned in the Sun code conventions (or if they are they're not where I expect them to be).

I'd still prefer it if _PostConstruct_ at least was at the top though - in most occurrences it seems to be near the top (_PreDestroy_ is all over the place lol).
I tend to count lifecycle methods in a similar vein to constructors, if anything a constructor is also a lifecycle method.
_PostConstruct_ I'd always expect to find at the top - when reading the code (and debugging) it means you don't start at the top, jump to the bottom, then go back to the top again.